### ISSUE TITLE
conv2d logical shape wrong padding

### ISSUE THREAD

**Issue by mert-kurttutan** [mert-kurttutan/issue-buffer#53]
In the following, logical output length is supposed to be 25 but it is 32 (equal to padded shape)
```python
# SPDX-FileCopyrightText: © 2026 Tenstorrent Inc.
#
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

import argparse

import torch

import ttnn


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Minimal reproduction for sharded Conv1d output-shape behavior.")
    parser.add_argument("--device-id", type=int, default=0, help="TT device id.")
    parser.add_argument("--l1-small-size", type=int, default=32768, help="CreateDevice l1_small_size.")
    parser.add_argument("--seed", type=int, default=0, help="Torch random seed.")
    parser.add_argument("--batch-size", type=int, default=1, help="Input batch size.")
    parser.add_argument("--input-length", type=int, default=50, help="Input sequence length.")
    parser.add_argument("--in-channels", type=int, default=512, help="Input channels.")
    parser.add_argument("--out-channels", type=int, default=512, help="Output channels.")
    parser.add_argument("--kernel-size", type=int, default=2, help="Conv1d kernel size.")
    parser.add_argument("--stride", type=int, default=2, help="Conv1d stride.")
    parser.add_argument("--padding", type=int, default=0, help="Conv1d padding.")
    parser.add_argument("--dilation", type=int, default=1, help="Conv1d dilation.")
    parser.add_argument("--groups", type=int, default=1, help="Conv1d groups.")
    parser.add_argument("--shard-height", type=int, default=32, help="Input shard height from the failing log.")
    parser.add_argument("--shard-width", type=int, default=512, help="Input shard width from the failing log.")
    parser.add_argument("--shard-cores-x", type=int, default=4, help="Input shard grid width from the failing log.")
    parser.add_argument("--shard-cores-y", type=int, default=1, help="Input shard grid height from the failing log.")
    return parser.parse_args()


def output_length_from_input_length(
    input_length: int,
    kernel_size: int,
    stride: int,
    padding: int,
    dilation: int,
) -> int:
    return (input_length + 2 * padding - dilation * (kernel_size - 1) - 1) // stride + 1


def get_shard_layout_for_conv(batch_size: int, input_length: int, in_channels: int):
    nhw = batch_size * input_length * in_channels
    if nhw >= 4 * in_channels:
        return ttnn.TensorMemoryLayout.HEIGHT_SHARDED
    if in_channels >= 4 * nhw:
        return ttnn.TensorMemoryLayout.WIDTH_SHARDED
    return ttnn.TensorMemoryLayout.BLOCK_SHARDED


def build_sharded_input_memory_config(args: argparse.Namespace) -> ttnn.MemoryConfig:
    return ttnn.create_sharded_memory_config(
        [args.shard_height, args.shard_width],
        core_grid=ttnn.CoreGrid(x=args.shard_cores_x, y=args.shard_cores_y),
        strategy=ttnn.ShardStrategy.HEIGHT,
        orientation=ttnn.ShardOrientation.ROW_MAJOR,
        use_height_and_width_as_shard_shape=True,
    )


def build_conv_config(args: argparse.Namespace) -> ttnn.Conv2dConfig:
    return ttnn.Conv2dConfig(
        weights_dtype=ttnn.bfloat8_b,
        shard_layout=get_shard_layout_for_conv(args.batch_size, args.input_length, args.in_channels),
        # output_layout=ttnn.TILE_LAYOUT,
        # deallocate_activation=False,
        # act_block_h_override=0,
        # act_block_w_div=1,
    )


def build_compute_config(device: ttnn.Device) -> ttnn.DeviceComputeKernelConfig:
    return ttnn.init_device_compute_kernel_config(
        device.arch(),
        # math_fidelity=ttnn.MathFidelity.LoFi,
        # math_approx_mode=True,
        # fp32_dest_acc_en=False,
        # packer_l1_acc=True,
    )


def build_tensors(
    args: argparse.Namespace,
    device: ttnn.Device,
    input_mem_config: ttnn.MemoryConfig,
) -> tuple[ttnn.Tensor, ttnn.Tensor, ttnn.Tensor, torch.Tensor, torch.nn.Conv1d]:
    torch_conv = torch.nn.Conv1d(
        in_channels=args.in_channels,
        out_channels=args.out_channels,
        kernel_size=args.kernel_size,
        stride=args.stride,
        padding=args.padding,
        dilation=args.dilation,
        groups=args.groups,
        bias=True,
    ).eval()

    torch_input = torch.randn(args.batch_size, args.input_length, args.in_channels, dtype=torch.float32)
    tt_input = ttnn.from_torch(
        torch_input.to(torch.bfloat16),
        dtype=ttnn.bfloat16,
        layout=ttnn.ROW_MAJOR_LAYOUT,
        device=device,
        memory_config=input_mem_config,
    )

    reshaped_weight = torch_conv.weight.reshape(
        args.out_channels,
        args.in_channels // args.groups,
        1,
        args.kernel_size,
    )
    tt_weight = ttnn.from_torch(reshaped_weight, dtype=ttnn.bfloat16)
    tt_bias = ttnn.from_torch(torch_conv.bias.reshape(1, 1, 1, -1), dtype=ttnn.bfloat16)

    return tt_input, tt_weight, tt_bias, torch_input, torch_conv


def print_tensor_info(prefix: str, tensor: ttnn.Tensor) -> None:
    print(
        f"{prefix}: shape={tensor.shape}, padded_shape={tensor.padded_shape}, "
        f"layout={tensor.layout}, memory_config={tensor.memory_config()}"
    )


def run_case(device: ttnn.Device, args: argparse.Namespace, case_name: str, input_mem_config: ttnn.MemoryConfig) -> None:
    conv_config = build_conv_config(args)
    compute_config = build_compute_config(device)
    expected_output_length = output_length_from_input_length(
        args.input_length,
        args.kernel_size,
        args.stride,
        args.padding,
        args.dilation,
    )

    tt_input, tt_weight, tt_bias, torch_input, torch_conv = build_tensors(args, device, input_mem_config)
    torch_ref = torch_conv(torch_input.permute(0, 2, 1).contiguous()).permute(0, 2, 1).contiguous()

    print("=" * 100)
    print(f"case={case_name}")
    print(f"expected_output_length={expected_output_length}, torch_ref_shape={tuple(torch_ref.shape)}")
    print_tensor_info("input", tt_input)

    try:
        out, [tt_weight, tt_bias] = ttnn.conv2d(
            input_tensor=ttnn.unsqueeze(tt_input, dim=1),
            weight_tensor=tt_weight,
            return_output_dim=False,
            return_weights_and_bias=True,
            device=device,
            in_channels=args.in_channels,
            out_channels=args.out_channels,
            batch_size=args.batch_size,
            input_height=1,
            input_width=args.input_length,
            kernel_size=[1, args.kernel_size],
            stride=[1, args.stride],
            padding=[0, 0, args.padding, args.padding],
            dilation=[1, args.dilation],
            groups=args.groups,
            bias_tensor=tt_bias,
            dtype=ttnn.bfloat16,
            conv_config=conv_config,
            compute_config=compute_config,
            slice_config=None,
        )
        print_tensor_info("out_before_squeeze", out)
        squeezed = ttnn.squeeze(out, dim=1)
        print_tensor_info("out_after_squeeze", squeezed)
        tt_output_torch = ttnn.to_torch(squeezed).to(torch.float32)
        print(f"tt_output_torch_shape={tuple(tt_output_torch.shape)}")
    finally:
        if "out" in locals():
            ttnn.deallocate(out)
        if "squeezed" in locals():
            ttnn.deallocate(squeezed)
        ttnn.deallocate(tt_input)


def main() -> None:
    args = parse_args()
    torch.manual_seed(args.seed)

    device = None
    try:
        device = ttnn.CreateDevice(device_id=args.device_id, l1_small_size=args.l1_small_size)
        if hasattr(device, "enable_program_cache"):
            device.enable_program_cache()

        sharded_input_memory_config = build_sharded_input_memory_config(args)
        run_case(device, args, case_name="interleaved_input", input_mem_config=ttnn.L1_MEMORY_CONFIG)
        run_case(device, args, case_name="height_sharded_input", input_mem_config=sharded_input_memory_config)
    finally:
        if device is not None:
            ttnn.close_device(device)


if __name__ == "__main__":
    main()

```

**Comment by mert-kurttutan** [comment]
This is due to input tensor being sharded tensor.
